model.py

```diff
@@ -9,43 +9,98 @@
+        self.rnn2 = nn.GRU(input_size=hidden_size+self.embedding_dim, hidden_size=hidden_size)
         self.linear = nn.Linear(in_features=hidden_size, out_features=output_size)
+        self.attention = GaussianAttention(query_size=hidden_size)
+
 
-    def forward(self, strokes):
+    def forward(self, strokes, texts, textsmask):
         hidden = None
-        output, hidden = self.step(strokes, hidden)
+        embedded_text = self.embedding(texts)
+        output, hidden = self.step(strokes, embedded_text, textsmask, hidden)
         return output
 
-    def step(self, x, hidden):
-        x, hidden = self.rnn(x, hidden)
-        output = self.linear(x)
-        return output, hidden
+    def step(self, x, embedded_text, textsmask, hidden):
+        if hidden is None:
+            hidden1 = None
+            previous_position = None
+            hidden2 = None
+        else:
+            hidden1, previous_position, hidden2 = hidden
+        o1, hidden1 = self.rnn1(x, hidden1)
+        context, previous_position = self.attention(o1, embedded_text, textsmask, previous_position)
+        o2, hidden2 = self.rnn2(torch.cat((o1, context), dim=-1), hidden2)
+        output = self.linear(o2)
+
+        result = output, (hidden1, previous_position, hidden2)
+        return result
 
     @property
     def device(self):
         return next(self.parameters()).device
 
     @argbind.bind
-    def sample(self, batch_size=1, num_steps=100, bias=0.0, stddev=0.1):
+    def sample(self, promptStr, num_steps=120, bias=0.0, stddev=0.1):
+        def sample_from_distribution(mean, std_logits):
+            std = torch.exp(std_logits - bias)
+            return torch.normal(mean=mean, std=std)
+
+        batch_size = 1
+        embedded_prompt = self.embedding(torch.tensor(self.dataset.text2code(promptStr), device=self.device).unsqueeze(dim=1))
+        prompt_mask = torch.ones((len(promptStr), 1), device=self.device)
         sample = torch.zeros(1, batch_size, 3, device=self.device)
         hidden = None
+        stddev = torch.log(torch.tensor([stddev, stddev]).view(1, batch_size, 2).to(self.device))
         output = []
-        for i in range(num_steps):
-            x, hidden = self.step(sample, hidden)
-            # Bias is weird now because we have a fixed stddev.  Instead of using bias, we could
-            # just adjust stddev.
-            #
-            # However, in the future, log-stddev will be part of the output of the model, and so it makes sense
-            # to apply bias independently.
-            zeros = torch.zeros(batch_size, 2, device=self.device)
-            std = torch.exp(torch.log(zeros + stddev) - bias)
-            noise = torch.normal(mean=zeros, std=std)
-            coords = x[:, :, 1:3] + noise
+        for _ in range(num_steps):
+            x, hidden = self.step(sample, embedded_prompt, prompt_mask, hidden)
+
+            mean = x[:, :, 1:3]
+            coords = sample_from_distribution(mean, stddev)
```

Comment:
No comments on this as we're going to abstract the sampling code to the output layer soon.